pkgs/os-specific/linux/nixos-rebuild/nixos-rebuild.sh

```diff
@@ -214,6 +214,49 @@ nixBuild() {
   fi
 }
 
+nixFlakeBuild() {
+    if [ -z "$buildHost" ]; then
+        nix build "$@" --out-link "${tmpDir}/result"
+        readlink -f "${tmpDir}/result"
+    else
+        local attr="$1"
+        shift 1
+        local evalArgs=()
+        local buildArgs=()
+        while [ "$#" -gt 0 ]; do
+            local i="$1"; shift 1
+            case "$i" in
+              --recreate-lock-file|--no-update-lock-file|--no-write-lock-file|--no-registries|--commit-lock-file)
+                evalArgs+=("$i")
+                ;;
+              --update-input)
+                local j="$1"; shift 1
+                evalArgs+=("$i" "$j")
+                ;;
+              --override-input)
+                local j="$1"; shift 1
+                local k="$1"; shift 1
+                evalArgs+=("$i" "$j" "$k")
+                ;;
+              *)
+                buildArgs+=("$i")
+                ;;
+            esac
+        done
+
+        local drv="$(nix "${flakeFlags[@]}" eval --raw "${attr}.drvPath" "${evalArgs[@]}" "${extraBuildArgs[@]}")"
+        if [ -a "$drv" ]; then
+            NIX_SSHOPTS=$SSHOPTS nix "${flakeFlags[@]}" copy --derivation --to "ssh://$buildHost" "$drv"
+            # The 'nix-command flakes' part in "${flakeFlags[@]}" is seen as two separate args over SSH
+            buildHostCmd nix --experimental-features "'nix-command flakes'" build "${buildArgs[@]}" --out-link "${tmpDir}/result" "$drv"
```

Comment:
This now requires that the remote host has `nixUnstable`, due to the `--experimental-features` flag. I think we should mention that somewhere.

Reply:
```
❯ ../nix/nixos-rebuild.sh switch --target-host monitor --flake ~/code/foo#monitor --use-remote-sudo
warning: Git tree '/home/asymmetric/code/foo' is dirty
building the system configuration...
warning: Git tree '/home/asymmetric/code/foo' is dirty
error: unrecognised flag '--experimental-features'
Try 'nix --help' for more information.
error: getting status of '/tmp/nixos-rebuild.u96Og6/result': No such file or directory
```

Reply:
Also:

```
+ ssh -o ControlMaster=auto -o ControlPath=/tmp/nixos-rebuild.fEtksJ/ssh-%n -o ControlPersist=60 monitor sudo -- nix-env -p /nix/var/nix/profiles/system --set /tmp/nixos-rebuild.fEtksJ/result
error: getting status of '/tmp/nixos-rebuild.fEtksJ/result': No such file or directory
+ cleanup
+ for ctrl in "$tmpDir"/ssh-*
+ ssh -o ControlPath=/tmp/nixos-rebuild.fEtksJ/ssh-monitor -O exit dummyhost
+ '[' -n monitor -a -n git+file:///home/asymmetric/code/foo ']'
+ buildHostCmd rm -rf /tmp/nixos-rebuild.fEtksJ
+ '[' -z monitor ']'
+ '[' -n '' ']'
+ ssh -o ControlMaster=auto -o ControlPath=/tmp/nixos-rebuild.fEtksJ/ssh-%n -o ControlPersist=60 monitor sudo -- rm -rf /tmp/nixos-rebuild.fEtksJ
```

Note the `ControlPath=/tmp/nixos-rebuild.fEtksJ/ssh-%n`, rather than `/tmp/nixos-rebuild.fEtksJ/ssh-monitor` . It asked for the SSH key's password again.

Reply:
I don't think that's true? As I understand it, `nixos-rebuild` uses it's own `nix` by default. I guess it could be documented for when `--no-build-nix`, `--build-host` and flakes are combined, but that seems like a rare use-case.

Reply:
Sorry, posted before I saw your later comments. I wonder if the `--use-remote-sudo` flag is responsible?

Reply:
There is already code that builds the latest Nix before attempting to rebuild to exactly avoid that situation, however, currently that is limited to the local machine.

https://github.com/NixOS/nixpkgs/blob/37f8212780cb360de213cfd7dda48c0c8fa80eab/pkgs/os-specific/linux/nixos-rebuild/nixos-rebuild.sh#L356-L392

Reply:
No, same thing:

```
❯ ../nix/nixos-rebuild.sh switch --target-host monitor --flake ~/code/foo#monitor
warning: Git tree '/home/asymmetric/code/foo' is dirty
building the system configuration...
warning: Git tree '/home/asymmetric/code/foo' is dirty
error: unrecognised flag '--experimental-features'
Try 'nix --help' for more information.
error: getting status of '/tmp/nixos-rebuild.Aa5unK/result': No such file or directory
```

Reply:
The documentation for the `--build-host` option says that it is also built remotely.

Reply:
Ah, the nix build only happens if flakes aren't used. That's annoying.

Reply:
I changed it to use old-style nix commands on the build host. It seemed like the most frictionless change to me.